library/swift/HeadersBuilder.swift

```diff
@@ -3,14 +3,19 @@ import Foundation
 private let kRestrictedPrefixes = [":", "x-envoy-mobile"]
 
 private func isRestrictedHeader(name: String) -> Bool {
-  return name == "host" || kRestrictedPrefixes.contains { name.hasPrefix($0) }
+  let isHostHeader = name.caseInsensitiveCompare("host") == .orderedSame
+  let hasRestrictedPrefix = kRestrictedPrefixes
+    .contains { name.range(of: $0, options: [.caseInsensitive, .anchored]) != nil }
+  return isHostHeader || hasRestrictedPrefix
 }
 
 /// Base builder class used to construct `Headers` instances.
+/// It preserves the original casing of headers and enforces
+/// a case-insensitive look up and setting of headers.
```

Comment:
Nit: lookup is a noun in this case

```suggestion
/// a case-insensitive lookup and setting of headers.
```